Removed line item limit and made sure the URL is shorter than 2083 ch…

peterfabian opened the pull request:
…aracters

- removed limit to 9 line items per Paypal order
- if the request URL with all line items will be longer than 2083 characters, send it as one line item
- fix the character limit function to account for URL encoding happening in http_build_query

fixes #19404 

Testing instructions:
Ideally, both following scenarios should work correctly:
- Order with 10 items should show all items correctly in PayPal (please note that products are sent as individual line items only if  either `wc_tax_enabled() == false` or `wc_prices_include_tax() == false`)
- Order with large number of items (20-30) should be sent to PayPal as one line item
- In either case, PayPal URL request should be shorter than 2083 characters.

peterfabian committed:
Removed line item limit and make sure the URL is shorter than 2083 characters

- removed limit to 9 line items per Paypal order
- if the request URL with all line items will be longer than 2083 characters, send it as one line item
- fix the character limit function to account for URL encoding happening in http_build_query

mikejolley commented on tests/unit-tests/gateways/paypal/request.php:
```diff
@@ -0,0 +1,261 @@
+		// wc_tax_enabled() === get_option( 'woocommerce_calc_taxes' ) === 'yes'
+		// wc_prices_include_tax() === wc_tax_enabled() && 'yes' === get_option( 'woocommerce_prices_include_tax' );.
+		$correct_options = array(
+			[ 'no', 'no', false ],
```
This syntax breaks under PHP 5.3/5.2 - Travis had some errors. Can you replace with standard `array` syntax?

mikejolley commented on tests/unit-tests/gateways/paypal/request.php:
```diff
@@ -0,0 +1,261 @@
+		// To test limit length properly, we need a product with a name that is shorter than 127 chars,
+		// but longer than 127 chars when URL-encoded.
+		if ( array_key_exists( 0, $this->products ) ) {
+			$this->products[0]->set_name( 'Dummy Product 😎😎😎😎😎😎😎😎😎😎😎' );
```
+1 experience point for emoji use.

mikejolley commented on includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php:
```diff
@@ -180,61 +191,93 @@ protected function get_shipping_args( $order ) {
 	 * @return array
 	 */
-	protected function get_line_item_args( $order ) {
+	protected function get_line_item_args( $order, $url_except_line_items_length ) {
```
If this is a new param we should set a default in case something is extending this class and not passing it.

mikejolley commented on includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php:
```diff
@@ -53,15 +53,16 @@ public function __construct( $gateway ) {
 	 */
 	public function get_request_url( $order, $sandbox = false ) {
+		if ( $sandbox ) {
+			$this->request_url_prefix = 'https://www.sandbox.paypal.com/cgi-bin/webscr?test_ipn=1&';
```
Just `endpoint` may be cleaner naming for this local param. Can we define it protected also?

mikejolley commented on includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php:
```diff
@@ -87,43 +92,49 @@ protected function limit_length( $string, $limit = 127 ) {
+				$initial_paypal_args,
+				$phone_number_args,
+				$shipping_args,
+				$this->get_line_item_args( $order, $url_except_line_items_length )
```
I wonder if instead of generating the args twice and passing sizes around we should instead change `get_line_item_args` to allow you to force it to a single item.

That way you can grab these args, and filter them (so extensions can make changes) THEN test length. If it's too long, call this and `get_line_item_args` again but this time force it to a single item?

peterfabian replied:
You're absolutely right, the length should be tested AFTER the filter, fixing it now.

However, I can't see if we can avoid running the filter and generating the args twice. 
Silly example: if the filter changed e.g one of shipping args based on number of line items, then we have to generate the args and run the filter again with one line item. 
Or am I missing something?

The change to make `get_line_item_args` accept the force single item parameter is much nicer, I completely agree, fixed that.

mikejolley commented on includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php:
```diff
@@ -364,7 +407,7 @@ protected function prepare_line_items( $order ) {
 		$index = ( count( $this->line_items ) / 4 ) + 1;
 
-		if ( $amount < 0 || $index > 9 ) {
+		if ( $amount < 0 ) {
```
Someone mentioned a 'max' amount in one of the issues too - shall we take care of that now also?

mikejolley replied:
If it cannot be taken care of easily, don't worry. It is edge case.

peterfabian replied:
I don't know if there is a good limit to use. The [wording on Paypal website](https://www.paypal.com/us/selfhelp/article/what-is-the-maximum-amount-i-can-send-with-my-paypal-account-faq732#personal) is a bit vague:
> If your account is Verified, there’s no limit on the total amount of money you can send from your account. You can send up to $60,000, but may be limited to $10,000, in a single transaction. (This amount can vary, though, depending on your currency.)

> If your account is Unverified, there’s a limit on the total amount of money you can send from your account.

Do you think we should add a limit of $10,000 based on this?

I actually think it's perhaps the Paypal's responsibility to tell you if you have sufficient funds/how much you can pay, so I don't think it belongs to WooCommerce core? That's a bit different from invalid request which can be too long or has too many fields.

mikejolley replied:
Lets leave this one then.

mikejolley commented on tests/unit-tests/gateways/paypal/request.php:
```diff
@@ -0,0 +1,261 @@
+	/**
+	 * Clean up order, deletes all products in order, too.
+	 */
+	protected function clean_up() {
```
I think you can add a teardown method rather than calling this manually.

claudiulodro replied:
`tearDown` is only called once at the end of each test method. Since all of the work in this is done in one test method, it wouldn't work.

mikejolley reviewed (changes requested):
Left some feedback and ideas :) Thanks!

@claudiulodro any feedback on the tests?

peterfabian committed:
Fixed compatibility with older PHP versions, removed superfluous asserts.

peterfabian committed:
Merge branch 'master' into fix/19404

peterfabian committed:
Test request length after applying filter

- added default value to get_line_item_args to not break interface
- smaller code review updates

peterfabian committed:
Fixed PHPCS violations

peterfabian committed:
Code duplication removed

Added test for non line item agrument presence
Fixed incorrect comment

peterfabian committed:
Fixed PHPCS violation

mikejolley commented on includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php:
```diff
@@ -72,12 +81,90 @@ public function get_request_url( $order, $sandbox = false ) {
+	protected function fix_request_length( $order, $paypal_args ) {
+		$max_paypal_length = 2083;
+		$query_candidate   = http_build_query( $paypal_args, '', '&' );
+		// If URL is longer than 2,083 chars, ignore line items and send cart to Paypal as a single item.
```
Why don't we add these inline comments to the main docblock instead? This function's purpose is to solve this issue.

mikejolley commented on includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php:
```diff
@@ -72,12 +81,90 @@ public function get_request_url( $order, $sandbox = false ) {
+	 *
+	 * @return array
+	 */
+	protected function get_non_line_item_args( $order ) {
```
This method name is bothering me for some reason :p Can we describe what this returns, rather than what it doesn't? transaction args?

peterfabian replied:
Good point, changed to `get_transaction_args`.

mikejolley commented on includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php:
```diff
@@ -72,12 +81,90 @@ public function get_request_url( $order, $sandbox = false ) {
+	 * Get args for paypal request, except for line item args.
+	 *
+	 * @param WC_Order $order Order object.
+	 *
```
No blank line needed here.

mikejolley commented on includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php:
```diff
@@ -72,12 +81,90 @@ public function get_request_url( $order, $sandbox = false ) {
+	 *
+	 * @param WC_Order $order Order to be sent to Paypal.
+	 * @param array    $paypal_args Arguments sent to Paypal in the request.
+	 *
```
We don't usually split params and return with a break.

mikejolley commented on includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php:
```diff
@@ -180,61 +236,87 @@ protected function get_shipping_args( $order ) {
 		if ( ( ! wc_tax_enabled() || ! wc_prices_include_tax() ) && $this->prepare_line_items( $order ) ) {
+			$line_item_args       = array();
+			$include_shipping_tax = false;
+			if ( $force_one_line_item ) {
```
Seems really odd to do this here and then repeat the single item code lower down. Why can't we check this on line 295?

peterfabian replied:
..because the `$include_shipping_tax` is different when generating single line item request in this branch and in the else branch for single line item.

The logic gets a bit convoluted here, but I understand it in the following way:
If we add `! $force_one_line_item` check to line 295, then we would need to check for `( ! wc_tax_enabled() || ! wc_prices_include_tax() ) && $this->prepare_line_items( $order )` again in the single line item branch to see if the shipping tax should be included, otherwise we would be changing the behaviour. 

I'm not 100% sure if this logic is correct (that inclusion of shipping tax depends on `wc_tax_enabled()`, `wc_prices_include_tax()` and also `prepare_line_items()`), maybe you can help me to find out/point me in the right direction, but I didn't want to change the logic.

To improve the code, I extracted the `include shipping tax` flag before the larger if statement, maybe it looks a bit better now?

peterfabian committed:
Line item logic refactored

Renamed function get_non_line_item_args
Documentation updates
...all part of code review fixes

peterfabian committed:
Removed incorrect comment

peterfabian committed:
Refactored line item logic

- prepare_line_items is not used as a test anymore, it only updates line items
- add_line_item is not used as a test anymore, only adds line items
- added method line_items_valid to test for valid line items
- shipping tax inclusion test moved to get_shipping_cost_line_item to encapsulate shipping tax-related logic

peterfabian commented:
Please don't merge yet, I'm testing some edge cases.

peterfabian committed:
Fixed shipping tax logic to behave the same as originally

peterfabian commented:
Ready for review, @mikejolley.

mikejolley reviewed (approved):
Much more readable and tests well. Thanks.
